baseplate/queue_consumer.py

```diff
@@ -0,0 +1,127 @@
+
+    :param baseplate.core.Baseplate baseplate: A baseplate instance for the
+        service.
+    :param kombu.Exchange exchange:
+    :param kombu.connection.Connection connection:
+    :param str queue_name: The name of the queue.
+    :param list routing_keys: List of routing keys.
+    :param handler: The handler method.
+
+    """
+    queues = []
+    for routing_key in routing_keys:
+        queue = Queue(
+            name=queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+        )
+        queues.append(queue)
+
+    print("registering %s as a handler for %r" % (handler.__name__, queues))
+    kombu_consumer = KombuConsumer(baseplate, connection, exchange, queues)
+    kombu_consumer.start()
+
+    print("waiting for messages")
+    while True:
+        with kombu_consumer.get_message() as (context, message, span):
+            handler(context, message.body, message)
+            message.ack()
+
+
+class _ConsumerWorker(ConsumerMixin):
+    def __init__(self, connection, queues, work_queue):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+
+    def get_consumers(self, Consumer, channel):
+        return [Consumer(
+            queues=self.queues,
+            on_message=self.on_message,
+        )]
+
+    def on_message(self, message):
+        self.work_queue.put(message)
+
+
+class ConsumerContext(object):
+    pass
+
+
+class KombuConsumer(object):
+    def __init__(self, baseplate, connection, exchange, queues):
+        self.baseplate = baseplate
+        self.connection = connection
+        self.exchange = exchange
+        self.queues = queues
+        self.work_queue = queue.Queue()
+        self.consumer = _ConsumerWorker(connection, queues, self.work_queue)
+        self.thread = None
+
+    def start(self):
+        assert self.thread is None, "consumer already started"
+        self.thread = Thread(target=self.consumer.run)
+        self.thread.name = "consumer message pump"
+        self.thread.daemon = True
+        self.thread.start()
+
+    @contextlib.contextmanager
+    def get_message(self, timeout=None):
+        with self.get_batch(max_items=1, timeout=timeout) as (context, batch, span):
+            yield context, batch[0], span
+
+    @contextlib.contextmanager
+    def get_batch(self, max_items, timeout):
+        assert self.thread is not None, "you must start the consumer first"
+
+        if timeout == 0:
+            block = False
+        else:
+            block = True
+
+        batch = []
+        for _ in range(max_items):
+            try:
+                item = self.work_queue.get(block=block, timeout=timeout)
+                batch.append(item)
+            except queue.Empty:
+                break
+
+        context = ConsumerContext()
+        with self.baseplate.make_server_span(context, self.queues[0].name) as span:
+            span.set_tag("message_bus.arrival", self.queues[0].routing_key)
```

Comment:
Should we add any other tags from https://github.com/opentracing/specification/blob/master/semantic_conventions.md#message-bus ?

Reply:
One thing I wondered about here: the list of queues probably all have different routing keys. It might be possible to figure out the actual routing key by looking at the message, but each message in the batch might have a different routing key. Maybe instead it makes more sense to just use the queue name here or omit entirely?

Reply:
Hrm, batches do sorta mess things up. I wonder if we should be specializing the single-item case better so it can have more detailed tags since it doesn't have to worry about this general case.

Probably easiest just to let the batch be tagged with some basic stuff like total number of items and the consumer name or something like that?